File: hashmap/380. Insert Delete GetRandom O(1).py

# ? https://leetcode.cn/problems/insert-delete-getrandom-o1/solution/o1-shi-jian-cha-ru-shan-chu-he-huo-qu-su-rlz2/
class RandomizedSet:

    def __init__(self):
        self.data = []
        self.map = {}

    def insert(self, val: int) -> bool:
        if val not in self.map:
            self.data.append(val)
            self.map[val] = len(self.data) - 1
            return True
        else:
            return False

    def remove(self, val: int) -> bool:
        if val not in self.map:
            return False
        else:
            eleIdx = self.map[val]
            lastEleInData = self.data[-1]
            self.data[eleIdx] = lastEleInData
            self.map[lastEleInData] = eleIdx
            self.data.pop()
            # * 从哈希表中删除的操作在最后进行的原因是防止要删除的是数组的「最后一个元素」，否则有前脚删完，后脚就又被 "self.map[lastEleInData] = eleIdx" 这个代码添加的风险
            del self.map[val]
            return True

    def getRandom(self) -> int:
        return choice(self.data)

File: hashmap/380. Insert Delete GetRandom O(1).js
// ? https://leetcode.cn/problems/insert-delete-getrandom-o1/solution/o1-shi-jian-cha-ru-shan-chu-he-huo-qu-su-rlz2/

var RandomizedSet = function () {
    this.data = [];
    this.map = new Map();
};

/**
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function (val) {
    if (this.map.has(val)) return false;

    let curLen = this.data.length;
    this.map.set(val, curLen);
    this.data.push(val);
    // console.log(this.data)
    // console.log(this.map)
    return true;
};

/**
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function (val) {
    if (!this.map.has(val)) return false;

    const delIdx = this.map.get(val); // 要删除的下标
    const lastIdx = this.data.length - 1; // 最后一个下标
    const lastVal = this.data[lastIdx]; // 最后一个值

    // 如果删除的不是最后一个元素，就把最后一个元素换到 delIdx
    if (delIdx !== lastIdx) {
        this.data[delIdx] = lastVal;
        this.map.set(lastVal, delIdx); // 更新 lastVal 在 map 中的新位置
    }

    // 删掉最后一个元素（原来的 lastVal 或目标值）
    this.data.pop();
    this.map.delete(val);

    return true;
};

/**
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function () {
    let randIdx = Math.floor(Math.random() * this.data.length);
    return this.data[randIdx];
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */
